File: programmers/lv2/기능 개발.py
"""
    lv.2 기능 개발
    
    1.  무조건 뒤의 작업은 앞의 작업이 완료되어야 같이 종료될 수 있다.
        따라서, 스택을 이용해 top이 100이상일때만 pop되도록 로직을 작성.
    2.  뽑는 방향을 두 리스트 모두 뒤집어 생각하면 편리.
"""


def solution(progresses, speeds):
    answer = []

    stack = list(reversed(progresses))
    speeds.reverse()

    while stack:
        for i in range(len(stack)):
            stack[i] += speeds[i]

        finished = 0

        while stack and stack[-1] >= 100:
            stack.pop()
            speeds.pop()

            finished += 1

        if finished > 0:
            answer.append(finished)

    return answer


result = solution([93, 30, 55], [1, 30, 5])

File: programmers/lv2/기능개발.js
/* 
    lv.2 기능개발 

    1.  날짜 1일 단위로 증가시키며 top의 값이 100이 될 때 100이상인 작업들을 모두 pop하면서 추가.
    2.  JS에서 별도로 스택이나 큐를 구현하지 않고 뒤집어 마지막 인덱스를 대상으로 스택 연산.
*/

function solution(progresses, speeds) {
  var answer = [];

  progresses.reverse();
  speeds.reverse();

  while (progresses.length !== 0) {
    progresses = progresses.map((element, idx) => {
      return (element += speeds[idx]);
    });

    let finished = 0;

    while (progresses.length !== 0 && progresses.at(-1) >= 100) {
      progresses.pop();
      speeds.pop();
      finished += 1;
    }

    if (finished > 0) {
      answer.push(finished);
    }
  }

  return answer;
}

const progresses = [95, 90, 99, 99, 80, 99];
const speeds = [1, 1, 1, 1, 1, 1];

console.log(solution(progresses, speeds));
